# 20 July 2024

## humann container build attempt

- <a href="https://apptainer.org/docs/user/main/docker_and_oci.html" target="_blank">Support for Docker and OCI Containers</a>

- <a href="https://hub.docker.com/r/biobakery/humann" target="_blank">biobakery/humann</a>

- <a href="https://hpcdocs.hpc.arizona.edu/events/workshop_materials/intro_to_containers/" target="_blank">Intro to Containers on HPC</a>

### definition/recipe file draft
``` 
Bootstrap: docker
From: ubuntu:18.04

%post
    # Update and install necessary packages
    apt-get update && apt-get install -y \
    python3 \
    python3-pip \
    bowtie2 \
    diamond \
    wget

    # Set installation directory
    export INSTALL_DIR=/xdisk/pi/member/tree-fruit/09humann

    # Install HUMAnN
    pip3 install humann

    # Download HUMAnN databases
    mkdir -p $INSTALL_DIR/databases
    humann_databases --download chocophlan full $INSTALL_DIR/databases
    humann_databases --download uniref uniref90_diamond $INSTALL_DIR/databases

    # Configure HUMAnN to use downloaded databases
    humann_config --update database_folders nucleotide $INSTALL_DIR/databases/chocophlan
    humann_config --update database_folders protein $INSTALL_DIR/databases/uniref

    # Clean up temporary files without attempting to remove system-protected directories
    apt-get clean
    rm -rf /var/lib/apt/lists/* /tmp/* || true
    rm -rf /var/tmp/* || true

%runscript
    # Default command to run HUMAnN
    echo "HUMAnN container is set up. Please provide input and output directories to run."

%environment
    # Set environment variables
    export INSTALL_DIR=/xdisk/pi/member/tree-fruit/09humann
    export PATH=$INSTALL_DIR/bin:$PATH

%labels
    MAINTAINER Name <email@example.com>
    DESCRIPTION "Container for HUMAnN bioinformatics tool"

%test
    # Run tests to ensure the installation works
    humann_test
```
### could not resolve permission errors
``` 
rm: cannot remove '/tmp/worker-63272-93100/trash': Permission denied
rm: cannot remove '/tmp/worker-63272-93118/trash': Permission denied
rm: cannot remove '/var/tmp/systemd-private-32e7500eecef4f8e982467c1423d89a9-chronyd.service-kP5gSi': Operation not permitted
rm: cannot remove '/var/tmp/systemd-private-32e7500eecef4f8e982467c1423d89a9-rtkit-daemon.service-oIeghW': Operation not permitted
FATAL:   While performing build: while running engine: exit status 1
```
***
## HPC Workshops and Training Materials
https://hpcdocs.hpc.arizona.edu/events/workshop_materials/
***
## Optimizing Lightweight File Processing for High-Performance Computing
**Objective**: Process files expeditiously in a lightweight environment while fully maintaining data integrity.

**Keywords**: Optimization, Efficiency, Memory

**Possible Code Techniques**:
- List Comprehensions
- Generator Expressions
- Tuple Unpacking
- Merge Dictionaries using Dictionary Unpacking
- `__slots__`
- `__str__` and `__repr__`
- `functools.lru_cache`
- `zip`
- `collections.ChainMap`
- `sys.exit()`
- `deque`
- `mmap`
- `array`
- f-Strings
- `if x in list`
- `str.join()`
- `str.split()`
- `set()`
- Walrus Operator (`:=`)
- Transforming Lists into a Dictionary
- List Slicing
- Garbage Collection Tuning
- Leverage Local Variable Scope
- Share Large Objects
- Cache and Reuse Objects
- `itertools`
***
## merge forward and reverse reads for humann entry with fastp

<a href="https://github.com/OpenGene/fastp/" target="_blank">fastp</a>

- Open the directory containing slurm and .py scripts in the terminal.

- Rid hidden and unexpected characters from files:
  
'sed -i 's/\r$//' fastp_merge.slurm'

'sed -i 's/\r$//' fastp_memory.py'

- Submit the script using the `sbatch` command:
'sbatch fastp_merge.slurm'
### fastp_merge.slurm script
``` slurm
#!/bin/bash
#SBATCH --job-name=fastp_merge
#SBATCH --account=kcooper
#SBATCH --partition=standard
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=35
#SBATCH --time=18:00:00
#SBATCH --mem-per-cpu=5gb
#SBATCH --mail-type=ALL

# Load the Python module
echo "Loading Python module"
module load python/3.8

# Directory containing the files
directory='/xdisk/kcooper/caparicio/tree-fruit/01d_fastp_reads_all'
output_dir="${directory}/output"
mkdir -p "${output_dir}"

# Run Python script with the necessary module
python3 fastp_memory.py "${directory}" "${output_dir}"
```
### fastp_memory.py script

- make .py script executable
  
'dir="/xdisk/kcooper/caparicio/tree-fruit/01d_fastp_reads_all/"'

'chmod +x "$dir"/*.py'

``` py
import glob
import os
import subprocess
import concurrent.futures
import sys

# Generator to yield pairs of files
def generate_file_pairs(directory):
    files_1 = sorted(glob.glob(os.path.join(directory, "*_1.trimmed.fastq")))
    files_2 = sorted(glob.glob(os.path.join(directory, "*_2.trimmed.fastq")))
    for file_1, file_2 in zip(files_1, files_2):
        yield (file_1, file_2)

# Function to merge two files using fastp
def merge_files(file_pair, output_dir):
    file_1, file_2 = file_pair
    output_file = os.path.join(output_dir, os.path.basename(file_1).replace("_1.trimmed.fastq", ".merged.fastq"))
    log_file = output_file.replace(".merged.fastq", ".log")
    command = [
        "apptainer", "exec", "https://depot.galaxyproject.org/singularity/fastp:0.23.3--h5f740d0_0",
        "fastp", "-i", file_1, "-I", file_2, "-o", output_file, "-m"
    ]
    with open(log_file, 'w') as log:
        subprocess.run(command, stdout=log, stderr=log)

def main(directory, output_dir):
    file_pairs = generate_file_pairs(directory)
    with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
        futures = [executor.submit(merge_files, pair, output_dir) for pair in file_pairs]
        for future in concurrent.futures.as_completed(futures):
            try:
                future.result()
            except Exception as e:
                print(f"Error merging files: {e}")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python3 fastp_memory.py <input_directory> <output_directory>")
        sys.exit(1)
    input_directory = sys.argv[1]
    output_directory = sys.argv[2]
    main(input_directory, output_directory)
```
***
## optimized metaspades - by each sample

- Open the directory containing slurm and .py scripts in the terminal.

- Rid hidden and unexpected characters from files:
  
'sed -i 's/\r$//' apple315_spade.slurm'

'sed -i 's/\r$//' apple315_spade.py'

- Submit the script using the `sbatch` command:
'sbatch apple315_spade.slurm'
### apple315_spade.slurm script
``` slurm
#!/bin/bash
#SBATCH --job-name=apple315_assembly
#SBATCH --account=kcooper
#SBATCH --partition=standard
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=38
#SBATCH --time=48:00:00
#SBATCH --mem-per-cpu=5gb    
#SBATCH --mail-type=ALL

module load spades
module load python/3.8

echo "Loading Python module"
cd /xdisk/kcooper/caparicio/tree-fruit/

# Run Python script with the necessary module
python3 /xdisk/kcooper/caparicio/tree-fruit/02batch/apple315_spade.py

NUM_THREADS=38
export OMP_NUM_THREADS=$NUM_THREADS

# Specify the files directly for apples315
f1="01c_fastp_reads/apples315_1.trimmed.fastq"
f2="01c_fastp_reads/apples315_2.trimmed.fastq"
sample_name="apples315"
output_dir="02_assembly/${sample_name}"

mkdir -p "$output_dir"
temp_dir="$output_dir/tmp" 
mkdir -p "$temp_dir"

if [ -f "${output_dir}/spades.log" ]; then
    echo "Resuming SPAdes assembly for $sample_name"
    spades.py --continue -o "$output_dir"
else
    echo "Starting new SPAdes assembly for $sample_name"
    spades.py --meta -k 21,33,55,77 -o "$output_dir" --tmp-dir "$temp_dir" -1 "$f1" -2 "$f2" --threads $NUM_THREADS
fi

if [ $? -eq 0 ]; then
    echo "Assembly completed successfully for $sample_name. Removing temporary files."
    rm -rf "$temp_dir"
else
    echo "Assembly incomplete or failed for $sample_name. Retaining files for troubleshooting."
fi
```
### apple315_spade.py script

- make .py script executable
  
'dir="/xdisk/kcooper/caparicio/tree-fruit/02batch/"'

'chmod +x "$dir"/*.py'

``` py
from functools import lru_cache
import mmap

@lru_cache(maxsize=None)
def process_file(file_path):
    with open(file_path, 'r') as f:
        # Memory-map the file, size 0 means whole file
        mmapped_file = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)
        for line in iter(mmapped_file.readline, b""):
            # Process each line
            pass
        mmapped_file.close()

def main():
    files = ["01c_fastp_reads/apples315_1.trimmed.fastq", "01c_fastp_reads/apples315_2.trimmed.fastq"]
    for file in files:
        process_file(file)

if __name__ == "__main__":
    main()
```
***
## humann python virtual environment install attempt

- <a href="https://hpcdocs.hpc.arizona.edu/software/popular_software/python/" target="_blank">Installing Python Packages Using a Virtual Environment
</a>

- <a href="https://github.com/biobakery/humann" target="_blank">humann</a>

- <a href="https://ua-researchcomputing-hpc.github.io/Python-Examples/Using-Virtualenv/" target="_blank">Using Virtual Environments Overview</a>

### install log
``` 
[caparicio@r4u13n1 ~]$ module load python/3.8
[caparicio@r4u13n1 ~]$ python3.8 -m venv /xdisk/kcooper/caparicio/virtual/humann38
[caparicio@r4u13n1 ~]$ source /xdisk/kcooper/caparicio/virtual/humann38/bin/activate
(humann38) [caparicio@r4u13n1 ~]$ pip install --upgrade pip
Requirement already satisfied: pip in /xdisk/kcooper/caparicio/virtual/humann38/lib/python3.8/site-packages (21.1.1)
Collecting pip
  Using cached pip-24.1.2-py3-none-any.whl (1.8 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 21.1.1
    Uninstalling pip-21.1.1:
      Successfully uninstalled pip-21.1.1
Successfully installed pip-24.1.2
(humann38) [caparicio@r4u13n1 ~]$ pip install humann
Collecting humann
  Using cached humann-3.9.tar.gz (107.2 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: humann
  Building wheel for humann (pyproject.toml) ... done
  Created wheel for humann: filename=humann-3.9-py3-none-any.whl size=105429071 sha256=083f25820038aa708de5ddbd588e67a8fb7692a7a21aaa9e3784976a6f24ec42
  Stored in directory: /home/u19/caparicio/.cache/pip/wheels/ba/0a/92/77b6b96fe7eb06c2e6dff0906d93fdfa0a347f780066c8e714
Successfully built humann
Installing collected packages: humann
Successfully installed humann-3.9


(humann) [caparicio@r4u13n1 ~]$ humann_databases --download chocophlan full /xdisk/kcooper/caparicio/tree-fruit/09humann
Creating subdirectory to install database: /xdisk/kcooper/caparicio/tree-fruit/09humann/chocophlan
Download URL: http://huttenhower.sph.harvard.edu/humann_data/chocophlan/full_chocophlan.v201901_v31.tar.gz
15.37 GB         
Extracting: /xdisk/kcooper/caparicio/tree-fruit/09humann/full_chocophlan.v201901_v31.tar.gz

Database installed: /xdisk/kcooper/caparicio/tree-fruit/09humann/chocophlan

HUMAnN configuration file updated: database_folders : nucleotide = /xdisk/kcooper/caparicio/tree-fruit/09humann/chocophlan
(humann) [caparicio@r4u13n1 ~]$ humann_databases --download uniref uniref90_diamond /xdisk/kcooper/caparicio/tree-fruit/09humann
Creating subdirectory to install database: /xdisk/kcooper/caparicio/tree-fruit/09humann/uniref
Download URL: http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_annotated/uniref90_annotated_v201901b_full.tar.gz
19.17 GB         
Extracting: /xdisk/kcooper/caparicio/tree-fruit/09humann/uniref90_annotated_v201901b_full.tar.gz

Database installed: /xdisk/kcooper/caparicio/tree-fruit/09humann/uniref

HUMAnN configuration file updated: database_folders : protein = /xdisk/kcooper/caparicio/tree-fruit/09humann/uniref
(humann) [caparicio@r4u13n1 ~]$ humann_databases --download uniref uniref50_diamond /xdisk/kcooper/caparicio/tree-fruit/09humann
Download URL: http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_annotated/uniref50_annotated_v201901b_full.tar.gz
6.43 GB         
Extracting: /xdisk/kcooper/caparicio/tree-fruit/09humann/uniref50_annotated_v201901b_full.tar.gz

Database installed: /xdisk/kcooper/caparicio/tree-fruit/09humann/uniref

HUMAnN configuration file updated: database_folders : protein = /xdisk/kcooper/caparicio/tree-fruit/09humann/uniref

(humann38) [caparicio@r4u10n2 ~]$ git clone https://github.com/biobakery/MetaPhlAn2.git
Cloning into 'MetaPhlAn2'...
remote: Enumerating objects: 135, done.
remote: Counting objects: 100% (135/135), done.
remote: Compressing objects: 100% (114/114), done.
remote: Total 135 (delta 11), reused 127 (delta 7), pack-reused 0
Receiving objects: 100% (135/135), 33.39 MiB | 30.16 MiB/s, done.
Resolving deltas: 100% (11/11), done.

[caparicio@r4u10n2 ~]$ module load python/3.8
[caparicio@r4u10n2 ~]$ source /xdisk/kcooper/caparicio/virtual/humann38/bin/activate
(humann38) [caparicio@r4u10n2 ~]$ git clone https://github.com/biobakery/MetaPhlAn2.git
Cloning into 'MetaPhlAn2'...
remote: Enumerating objects: 135, done.
remote: Counting objects: 100% (135/135), done.
remote: Compressing objects: 100% (114/114), done.
remote: Total 135 (delta 11), reused 127 (delta 7), pack-reused 0
Receiving objects: 100% (135/135), 33.39 MiB | 30.16 MiB/s, done.
Resolving deltas: 100% (11/11), done.
(humann38) [caparicio@r4u10n2 ~]$ metaphlan2 --version
-bash: metaphlan2: command not found
(humann38) [caparicio@r4u10n2 ~]$ cd /xdisk/kcooper/caparicio/virtual/humann38/bin
(humann38) [caparicio@r4u10n2 bin]$ pwd
/xdisk/kcooper/caparicio/virtual/humann38/bin
(humann38) [caparicio@r4u10n2 bin]$ git clone https://github.com/biobakery/MetaPhlAn2.git
Cloning into 'MetaPhlAn2'...
remote: Enumerating objects: 135, done.
remote: Counting objects: 100% (135/135), done.
remote: Compressing objects: 100% (114/114), done.
remote: Total 135 (delta 11), reused 127 (delta 7), pack-reused 0
Receiving objects: 100% (135/135), 33.39 MiB | 32.26 MiB/s, done.
Resolving deltas: 100% (11/11), done.
(humann38) [caparicio@r4u10n2 bin]$ cd MetaPhlAn2
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ pip install .
ERROR: Directory '.' is not installable. Neither 'setup.py' nor 'pyproject.toml' found.
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ python setup.py install
python: can't open file 'setup.py': [Errno 2] No such file or directory
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ cmd
-bash: cmd: command not found
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ pwd
/xdisk/kcooper/caparicio/virtual/humann38/bin/MetaPhlAn2
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ $ export PATH=`pwd`:$PATH
-bash: $: command not found
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ export PATH=`pwd`:$PATH
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ export mpa_dir=`pwd`
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ metaphlan2.py --version
Error! numpy python library not detected!!
(humann38) [caparicio@r4u10n2 MetaPhlAn2]$ cd ..
(humann38) [caparicio@r4u10n2 bin]$ pwd
/xdisk/kcooper/caparicio/virtual/humann38/bin
(humann38) [caparicio@r4u10n2 bin]$ pip install numpy
Collecting numpy
  Downloading numpy-1.24.4-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.6 kB)
Downloading numpy-1.24.4-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.3/17.3 MB 40.1 MB/s eta 0:00:00
Installing collected packages: numpy
Successfully installed numpy-1.24.4
(humann38) [caparicio@r4u10n2 bin]$ metaphlan2.py --version
Warning! Biom python library not detected!
 Exporting to biom format will not work!
MetaPhlAn version 2.6.0	(19 August 2016)
(humann38) [caparicio@r4u10n2 bin]$ ^C
(humann38) [caparicio@r4u10n2 bin]$ pip install biom-format
Collecting biom-format
  Using cached biom_format-2.1.16-cp38-cp38-linux_x86_64.whl
Collecting click (from biom-format)
  Downloading click-8.1.7-py3-none-any.whl.metadata (3.0 kB)
Requirement already satisfied: numpy>=1.9.2 in /xdisk/kcooper/caparicio/virtual/humann38/lib/python3.8/site-packages (from biom-format) (1.24.4)
Collecting scipy>=1.3.1 (from biom-format)
  Downloading scipy-1.10.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (58 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.9/58.9 kB 1.2 MB/s eta 0:00:00
Collecting pandas>=0.20.0 (from biom-format)
  Downloading pandas-2.0.3-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (18 kB)
Collecting h5py (from biom-format)
  Downloading h5py-3.11.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.5 kB)
Collecting python-dateutil>=2.8.2 (from pandas>=0.20.0->biom-format)
  Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting pytz>=2020.1 (from pandas>=0.20.0->biom-format)
  Downloading pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)
Collecting tzdata>=2022.1 (from pandas>=0.20.0->biom-format)
  Downloading tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)
Collecting six>=1.5 (from python-dateutil>=2.8.2->pandas>=0.20.0->biom-format)
  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Downloading pandas-2.0.3-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.4 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 50.7 MB/s eta 0:00:00
Downloading scipy-1.10.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.5 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.5/34.5 MB 59.3 MB/s eta 0:00:00
Downloading click-8.1.7-py3-none-any.whl (97 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.9/97.9 kB 4.0 MB/s eta 0:00:00
Downloading h5py-3.11.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.3/5.3 MB 18.7 MB/s eta 0:00:00
Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
Downloading pytz-2024.1-py2.py3-none-any.whl (505 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 505.5/505.5 kB 18.1 MB/s eta 0:00:00
Downloading tzdata-2024.1-py2.py3-none-any.whl (345 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.4/345.4 kB 15.4 MB/s eta 0:00:00
Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Installing collected packages: pytz, tzdata, six, scipy, h5py, click, python-dateutil, pandas, biom-format
Successfully installed biom-format-2.1.16 click-8.1.7 h5py-3.11.0 pandas-2.0.3 python-dateutil-2.9.0.post0 pytz-2024.1 scipy-1.10.1 six-1.16.0 tzdata-2024.1
(humann38) [caparicio@r4u10n2 bin]$ metaphlan2.py --version
MetaPhlAn version 2.6.0	(19 August 2016)
(humann38) [caparicio@r4u10n2 bin]$ python -c "import numpy; print(numpy.__version__)"
1.24.4
(humann38) [caparicio@r4u10n2 bin]$ python -c "import biom; print(biom.__version__)"
2.1.16
```
### humann step 1 slurm script - attempt 1
``` 
HUMANN_STEP1.SLURM
#!/bin/bash
#SBATCH --job-name=humann_step1
#SBATCH --account=pi
#SBATCH --partition=standard
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=35
#SBATCH --time=18:00:00
#SBATCH --mem-per-cpu=5gb
#SBATCH --mail-type=ALL

# Load the necessary modules
source /xdisk/pi/member/virtual/humann/bin/activate
module load python/3.8

# Set the directory containing the input files and the output directory
input_directory='/xdisk/pi/member/tree-fruit/01d_fastp_reads_all/output'
output_directory='/xdisk/pi/member/tree-fruit/09humann/output'
chocophlan_database='/xdisk/pi/member/tree-fruit/09humann/chocophlan'

# Run the Python script
python3 humann_step1.py "${input_directory}" "${output_directory}" "${chocophlan_database}"
```
### humann step 1 .py script - attempt 1
``` 
#!/usr/bin/env python3

import os
import sys
import subprocess
from concurrent.futures import ThreadPoolExecutor

def generate_file_list(input_directory):
    """Generate a list of .fastq files in the input directory."""
    return [os.path.join(input_directory, file) for file in os.listdir(input_directory) if file.endswith('.fastq')]

def run_humann(sample, output_dir, chocophlan_database, uniref_database):
    """Run HUMAnN for a given sample."""
    output_sample_dir = os.path.join(output_dir, os.path.basename(sample).replace('.fastq', ''))
    os.makedirs(output_sample_dir, exist_ok=True)
    command = humann --input {sample} --output {output_sample_dir} --input_format fastq --nucleotide-database {chocophlan_database} --protein-database {uniref_database}"
    subprocess.run(command, shell=True, check=True)

def main(input_directory, output_directory, chocophlan_database, uniref_database):
    samples = generate_file_list(input_directory)
    with ThreadPoolExecutor(max_workers=35) as executor:
        futures = [executor.submit(run_humann, sample, output_directory, chocophlan_database, uniref_database) for sample in samples]
        for future in futures:
            future.result()

if __name__ == "__main__":
    if len(sys.argv) != 5:
        print("Usage: python humann_step1.py <input_directory> <output_directory> <chocophlan_database> <uniref_database>")
        sys.exit(1)
    input_directory = sys.argv[1]
    output_directory = sys.argv[2]
    chocophlan_database = sys.argv[3]
    uniref_database = sys.argv[4]
    main(input_directory, output_directory, chocophlan_database, uniref_database)
```
### humann step 1 slurm job log - error 1    
```
command = humann --input {sample} --output {output_sample_dir} --input_format fastq --nucleotide-database {chocophlan_database} --protein-database {uniref_database}"
                             ^
SyntaxError: invalid syntax
```
### debug attempt 1 - replace .py code section:

```python
command = humann --input {sample} --output {output_sample_dir} --input_format fastq --nucleotide-database {chocophlan_database} --protein-database {uniref_database}"
```
with 

```python
command = f"humann --input {sample} --output {output_sample_dir} --input_format fastq --nucleotide-database {chocophlan_database} --protein-database {uniref_database}"
```
### humann step 1 slurm job log - error 2
```
/xdisk/pi/member/virtual/humann/bin/python: error while loading shared libraries: libpython3.11.so.1.0: cannot open shared object file: No such file or directory
Traceback (most recent call last):
  File "humann_step1.py", line 41, in <module>
    main(input_directory, output_directory, chocophlan_database, uniref_database)
  File "humann_step1.py", line 31, in main
    future.result()
  File "/opt/ohpc/pub/apps/python/3.8.12/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/opt/ohpc/pub/apps/python/3.8.12/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/opt/ohpc/pub/apps/python/3.8.12/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "humann_step1.py", line 24, in run_humann
    subprocess.run(command, shell=True, check=True)
  File "/opt/ohpc/pub/apps/python/3.8.12/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'humann --input /xdisk/pi/member/tree-fruit/01d_fastp_reads_all/output/peaches273.merged.fastq --output /xdisk/pi/member/tree-fruit/09humann/output/peaches273.merged --input_format fastq --nucleotide-database /xdisk/pi/member/tree-fruit/09humann/chocophlan --protein-database /xdisk/pi/member/tree-fruit/09humann/uniref90' returned non-zero exit status 127.
```
numpy and metaphlan not found - DELETE INSTALL
***
## mamba reinstall
- <a href="https://github.com/conda-forge/miniforge" target="_blank">miniforge</a>
- mamba mistakenly deleted in attempt to make room for installs in home drive
- was directed to 'miniforge' when searching for 'mamba'
```
$ wget "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh"
--2024-07-26 12:57:26--  https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh
Resolving github.com (github.com)... 140.82.116.3
Connecting to github.com (github.com)|140.82.116.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github.com/conda-forge/miniforge/releases/download/24.3.0-0/Miniforge3-Linux-x86_64.sh [following]
--2024-07-26 12:57:26--  https://github.com/conda-forge/miniforge/releases/download/24.3.0-0/Miniforge3-Linux-x86_64.sh
Reusing existing connection to github.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/221584272/44f31adc-7a87-4d05-8e42-82d599d5b648?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240726%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240726T195726Z&X-Amz-Expires=300&X-Amz-Signature=c387f39dac2890e9cd087ad050c266ae2916634fba20ecdc50237beae023f3d1&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=221584272&response-content-disposition=attachment%3B%20filename%3DMiniforge3-Linux-x86_64.sh&response-content-type=application%2Foctet-stream [following]
--2024-07-26 12:57:27--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/221584272/44f31adc-7a87-4d05-8e42-82d599d5b648?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240726%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240726T195726Z&X-Amz-Expires=300&X-Amz-Signature=c387f39dac2890e9cd087ad050c266ae2916634fba20ecdc50237beae023f3d1&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=221584272&response-content-disposition=attachment%3B%20filename%3DMiniforge3-Linux-x86_64.sh&response-content-type=application%2Foctet-stream
Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 82796318 (79M) [application/octet-stream]
Saving to: 'Miniforge3-Linux-x86_64.sh.1'

100%[======================================>] 82,796,318   105MB/s   in 0.8s   

2024-07-26 12:57:28 (105 MB/s) - 'Miniforge3-Linux-x86_64.sh.1' saved [82796318/82796318]

[caparicio@r3u32n2 ~]$ bash Miniforge3-$(uname)-$(uname -m).sh

Welcome to Miniforge3 24.3.0-0

In order to continue the installation process, please review the license
agreement.
Please, press ENTER to continue
>>> 
Miniforge installer code uses BSD-3-Clause license as stated below.

Binary packages that come with it have their own licensing terms
and by installing miniforge you agree to the licensing terms of individual
packages as well. They include different OSI-approved licenses including
the GNU General Public License and can be found in pkgs/<pkg-name>/info/licenses
folders.

Miniforge installer comes with a boostrapping executable that is used
when installing miniforge and is deleted after miniforge is installed.
The bootstrapping executable uses micromamba, cli11, cpp-filesystem,
curl, c-ares, krb5, libarchive, libev, lz4, nghttp2, openssl, libsolv,
nlohmann-json, reproc and zstd which are licensed under BSD-3-Clause,
MIT and OpenSSL licenses. Licenses and copyright notices of these
projects can be found at the following URL.
https://github.com/conda-forge/micromamba-feedstock/tree/master/recipe.

=============================================================================

Copyright (c) 2019-2022, conda-forge
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its contributors
may be used to endorse or promote products derived from this software without
specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Do you accept the license terms? [yes|no]
>>> 
Please answer 'yes' or 'no':'
>>> yes

Miniforge3 will now be installed into this location:
/home/u19/caparicio/miniforge3

  - Press ENTER to confirm the location
  - Press CTRL-C to abort the installation
  - Or specify a different location below

[/home/u19/caparicio/miniforge3] >>> 
PREFIX=/home/u19/caparicio/miniforge3
Unpacking payload ...
Extracting _libgcc_mutex-0.1-conda_forge.tar.bz2
Extracting ca-certificates-2024.2.2-hbcca054_0.conda
Extracting ld_impl_linux-64-2.40-h41732ed_0.conda
Extracting libstdcxx-ng-13.2.0-h7e041cc_5.conda
Extracting pybind11-abi-4-hd8ed1ab_3.tar.bz2
Extracting python_abi-3.10-4_cp310.conda
Extracting tzdata-2024a-h0c530f3_0.conda
Extracting libgomp-13.2.0-h807b86a_5.conda
Extracting _openmp_mutex-4.5-2_gnu.tar.bz2
Extracting libgcc-ng-13.2.0-h807b86a_5.conda
Extracting bzip2-1.0.8-hd590300_5.conda
Extracting c-ares-1.28.1-hd590300_0.conda
Extracting fmt-10.2.1-h00ab1b0_0.conda
Extracting icu-73.2-h59595ed_0.conda
Extracting keyutils-1.6.1-h166bdaf_0.tar.bz2
Extracting libev-4.33-hd590300_2.conda
Extracting libffi-3.4.2-h7f98852_5.tar.bz2
Extracting libiconv-1.17-hd590300_2.conda
Extracting libnsl-2.0.1-hd590300_0.conda
Extracting libuuid-2.38.1-h0b41bf4_0.conda
Extracting libxcrypt-4.4.36-hd590300_1.conda
Extracting libzlib-1.2.13-hd590300_5.conda
Extracting lz4-c-1.9.4-hcb278e6_0.conda
Extracting lzo-2.10-h516909a_1000.tar.bz2
Extracting ncurses-6.4.20240210-h59595ed_0.conda
Extracting openssl-3.2.1-hd590300_1.conda
Extracting reproc-14.2.4.post0-hd590300_1.conda
Extracting xz-5.2.6-h166bdaf_0.tar.bz2
Extracting yaml-cpp-0.8.0-h59595ed_0.conda
Extracting libedit-3.1.20191231-he28a2e2_2.tar.bz2
Extracting libnghttp2-1.58.0-h47da74e_1.conda
Extracting libsolv-0.7.28-hfc55251_2.conda
Extracting libsqlite-3.45.2-h2797004_0.conda
Extracting libssh2-1.11.0-h0841786_0.conda
Extracting libxml2-2.12.6-h232c23b_1.conda
Extracting readline-8.2-h8228510_1.conda
Extracting reproc-cpp-14.2.4.post0-h59595ed_1.conda
Extracting tk-8.6.13-noxft_h4845f30_101.conda
Extracting zstd-1.5.5-hfc55251_0.conda
Extracting krb5-1.21.2-h659d440_0.conda
Extracting libarchive-3.7.2-h2aa1ff5_1.conda
Extracting python-3.10.14-hd12c33a_0_cpython.conda
Extracting libcurl-8.7.1-hca28451_0.conda
Extracting menuinst-2.0.2-py310hff52083_0.conda
Extracting archspec-0.2.3-pyhd8ed1ab_0.conda
Extracting boltons-24.0.0-pyhd8ed1ab_0.conda
Extracting brotli-python-1.1.0-py310hc6cd4ac_1.conda
Extracting certifi-2024.2.2-pyhd8ed1ab_0.conda
Extracting charset-normalizer-3.3.2-pyhd8ed1ab_0.conda
Extracting colorama-0.4.6-pyhd8ed1ab_0.tar.bz2
Extracting distro-1.9.0-pyhd8ed1ab_0.conda
Extracting idna-3.6-pyhd8ed1ab_0.conda
Extracting jsonpointer-2.4-py310hff52083_3.conda
Extracting libmamba-1.5.8-had39da4_0.conda
Extracting packaging-24.0-pyhd8ed1ab_0.conda
Extracting platformdirs-4.2.0-pyhd8ed1ab_0.conda
Extracting pluggy-1.4.0-pyhd8ed1ab_0.conda
Extracting pycosat-0.6.6-py310h2372a71_0.conda
Extracting pycparser-2.22-pyhd8ed1ab_0.conda
Extracting pysocks-1.7.1-pyha2e5f31_6.tar.bz2
Extracting ruamel.yaml.clib-0.2.8-py310h2372a71_0.conda
Extracting setuptools-69.5.1-pyhd8ed1ab_0.conda
Extracting truststore-0.8.0-pyhd8ed1ab_0.conda
Extracting wheel-0.43.0-pyhd8ed1ab_1.conda
Extracting cffi-1.16.0-py310h2fee648_0.conda
Extracting jsonpatch-1.33-pyhd8ed1ab_0.conda
Extracting libmambapy-1.5.8-py310h39ff949_0.conda
Extracting pip-24.0-pyhd8ed1ab_0.conda
Extracting ruamel.yaml-0.18.6-py310h2372a71_0.conda
Extracting tqdm-4.66.2-pyhd8ed1ab_0.conda
Extracting urllib3-2.2.1-pyhd8ed1ab_0.conda
Extracting requests-2.31.0-pyhd8ed1ab_0.conda
Extracting zstandard-0.22.0-py310h1275a96_0.conda
Extracting conda-package-streaming-0.9.0-pyhd8ed1ab_0.conda
Extracting conda-package-handling-2.2.0-pyh38be061_0.conda
Extracting conda-24.3.0-py310hff52083_0.conda
Extracting conda-libmamba-solver-24.1.0-pyhd8ed1ab_0.conda
Extracting mamba-1.5.8-py310h51d5547_0.conda

Installing base environment...

Transaction

  Prefix: /home/u19/caparicio/miniforge3

  Updating specs:

   - conda-forge/linux-64::_libgcc_mutex==0.1=conda_forge[md5=d7c89558ba9fa0495403155b64376d81]
   - conda-forge/linux-64::ca-certificates==2024.2.2=hbcca054_0[md5=2f4327a1cbe7f022401b236e915a5fef]
   - conda-forge/linux-64::ld_impl_linux-64==2.40=h41732ed_0[md5=7aca3059a1729aa76c597603f10b0dd3]
   - conda-forge/linux-64::libstdcxx-ng==13.2.0=h7e041cc_5[md5=f6f6600d18a4047b54f803cf708b868a]
   - conda-forge/noarch::pybind11-abi==4=hd8ed1ab_3[md5=878f923dd6acc8aeb47a75da6c4098be]
   - conda-forge/linux-64::python_abi==3.10=4_cp310[md5=26322ec5d7712c3ded99dd656142b8ce]
   - conda-forge/noarch::tzdata==2024a=h0c530f3_0[md5=161081fc7cec0bfda0d86d7cb595f8d8]
   - conda-forge/linux-64::libgomp==13.2.0=h807b86a_5[md5=d211c42b9ce49aee3734fdc828731689]
   - conda-forge/linux-64::_openmp_mutex==4.5=2_gnu[md5=73aaf86a425cc6e73fcf236a5a46396d]
   - conda-forge/linux-64::libgcc-ng==13.2.0=h807b86a_5[md5=d4ff227c46917d3b4565302a2bbb276b]
   - conda-forge/linux-64::bzip2==1.0.8=hd590300_5[md5=69b8b6202a07720f448be700e300ccf4]
   - conda-forge/linux-64::c-ares==1.28.1=hd590300_0[md5=dcde58ff9a1f30b0037a2315d1846d1f]
   - conda-forge/linux-64::fmt==10.2.1=h00ab1b0_0[md5=35ef8bc24bd34074ebae3c943d551728]
   - conda-forge/linux-64::icu==73.2=h59595ed_0[md5=cc47e1facc155f91abd89b11e48e72ff]
   - conda-forge/linux-64::keyutils==1.6.1=h166bdaf_0[md5=30186d27e2c9fa62b45fb1476b7200e3]
   - conda-forge/linux-64::libev==4.33=hd590300_2[md5=172bf1cd1ff8629f2b1179945ed45055]
   - conda-forge/linux-64::libffi==3.4.2=h7f98852_5[md5=d645c6d2ac96843a2bfaccd2d62b3ac3]
   - conda-forge/linux-64::libiconv==1.17=hd590300_2[md5=d66573916ffcf376178462f1b61c941e]
   - conda-forge/linux-64::libnsl==2.0.1=hd590300_0[md5=30fd6e37fe21f86f4bd26d6ee73eeec7]
   - conda-forge/linux-64::libuuid==2.38.1=h0b41bf4_0[md5=40b61aab5c7ba9ff276c41cfffe6b80b]
   - conda-forge/linux-64::libxcrypt==4.4.36=hd590300_1[md5=5aa797f8787fe7a17d1b0821485b5adc]
   - conda-forge/linux-64::libzlib==1.2.13=hd590300_5[md5=f36c115f1ee199da648e0597ec2047ad]
   - conda-forge/linux-64::lz4-c==1.9.4=hcb278e6_0[md5=318b08df404f9c9be5712aaa5a6f0bb0]
   - conda-forge/linux-64::lzo==2.10=h516909a_1000[md5=bb14fcb13341b81d5eb386423b9d2bac]
   - conda-forge/linux-64::ncurses==6.4.20240210=h59595ed_0[md5=97da8860a0da5413c7c98a3b3838a645]
   - conda-forge/linux-64::openssl==3.2.1=hd590300_1[md5=9d731343cff6ee2e5a25c4a091bf8e2a]
   - conda-forge/linux-64::reproc==14.2.4.post0=hd590300_1[md5=82ca53502dfd5a64a80dee76dae14685]
   - conda-forge/linux-64::xz==5.2.6=h166bdaf_0[md5=2161070d867d1b1204ea749c8eec4ef0]
   - conda-forge/linux-64::yaml-cpp==0.8.0=h59595ed_0[md5=965eaacd7c18eb8361fd12bb9e7a57d7]
   - conda-forge/linux-64::libedit==3.1.20191231=he28a2e2_2[md5=4d331e44109e3f0e19b4cb8f9b82f3e1]
   - conda-forge/linux-64::libnghttp2==1.58.0=h47da74e_1[md5=700ac6ea6d53d5510591c4344d5c989a]
   - conda-forge/linux-64::libsolv==0.7.28=hfc55251_2[md5=535bafe1ed0a5bdd3f4c125ca05d378c]
   - conda-forge/linux-64::libsqlite==3.45.2=h2797004_0[md5=866983a220e27a80cb75e85cb30466a1]
   - conda-forge/linux-64::libssh2==1.11.0=h0841786_0[md5=1f5a58e686b13bcfde88b93f547d23fe]
   - conda-forge/linux-64::libxml2==2.12.6=h232c23b_1[md5=6853448e9ca1cfd5f15382afd2a6d123]
   - conda-forge/linux-64::readline==8.2=h8228510_1[md5=47d31b792659ce70f470b5c82fdfb7a4]
   - conda-forge/linux-64::reproc-cpp==14.2.4.post0=h59595ed_1[md5=715e1d720ec1a03715bebd237972fca5]
   - conda-forge/linux-64::tk==8.6.13=noxft_h4845f30_101[md5=d453b98d9c83e71da0741bb0ff4d76bc]
   - conda-forge/linux-64::zstd==1.5.5=hfc55251_0[md5=04b88013080254850d6c01ed54810589]
   - conda-forge/linux-64::krb5==1.21.2=h659d440_0[md5=cd95826dbd331ed1be26bdf401432844]
   - conda-forge/linux-64::libarchive==3.7.2=h2aa1ff5_1[md5=3bf887827d1968275978361a6e405e4f]
   - conda-forge/linux-64::python==3.10.14=hd12c33a_0_cpython[md5=2b4ba962994e8bd4be9ff5b64b75aff2]
   - conda-forge/linux-64::libcurl==8.7.1=hca28451_0[md5=755c7f876815003337d2c61ff5d047e5]
   - conda-forge/linux-64::menuinst==2.0.2=py310hff52083_0[md5=4837faab0d3e665df57fef662148c6a3]
   - conda-forge/noarch::archspec==0.2.3=pyhd8ed1ab_0[md5=192278292e20704f663b9c766909d67b]
   - conda-forge/noarch::boltons==24.0.0=pyhd8ed1ab_0[md5=61de176bd62041f9cd5bd4fcd09eb0ff]
   - conda-forge/linux-64::brotli-python==1.1.0=py310hc6cd4ac_1[md5=1f95722c94f00b69af69a066c7433714]
   - conda-forge/noarch::certifi==2024.2.2=pyhd8ed1ab_0[md5=0876280e409658fc6f9e75d035960333]
   - conda-forge/noarch::charset-normalizer==3.3.2=pyhd8ed1ab_0[md5=7f4a9e3fcff3f6356ae99244a014da6a]
   - conda-forge/noarch::colorama==0.4.6=pyhd8ed1ab_0[md5=3faab06a954c2a04039983f2c4a50d99]
   - conda-forge/noarch::distro==1.9.0=pyhd8ed1ab_0[md5=bbdb409974cd6cb30071b1d978302726]
   - conda-forge/noarch::idna==3.6=pyhd8ed1ab_0[md5=1a76f09108576397c41c0b0c5bd84134]
   - conda-forge/linux-64::jsonpointer==2.4=py310hff52083_3[md5=08ec1463dbc5c806a32fc431874032ca]
   - conda-forge/linux-64::libmamba==1.5.8=had39da4_0[md5=def669885dc103d8acb7ac2ac35e0b2f]
   - conda-forge/noarch::packaging==24.0=pyhd8ed1ab_0[md5=248f521b64ce055e7feae3105e7abeb8]
   - conda-forge/noarch::platformdirs==4.2.0=pyhd8ed1ab_0[md5=a0bc3eec34b0fab84be6b2da94e98e20]
   - conda-forge/noarch::pluggy==1.4.0=pyhd8ed1ab_0[md5=139e9feb65187e916162917bb2484976]
   - conda-forge/linux-64::pycosat==0.6.6=py310h2372a71_0[md5=0adaac9a86d59adae2bc86b3cdef2df1]
   - conda-forge/noarch::pycparser==2.22=pyhd8ed1ab_0[md5=844d9eb3b43095b031874477f7d70088]
   - conda-forge/noarch::pysocks==1.7.1=pyha2e5f31_6[md5=2a7de29fb590ca14b5243c4c812c8025]
   - conda-forge/linux-64::ruamel.yaml.clib==0.2.8=py310h2372a71_0[md5=dcf6d2535586c77b31425ed835610c54]
   - conda-forge/noarch::setuptools==69.5.1=pyhd8ed1ab_0[md5=7462280d81f639363e6e63c81276bd9e]
   - conda-forge/noarch::truststore==0.8.0=pyhd8ed1ab_0[md5=08316d001eca8854392cf2837828ea11]
   - conda-forge/noarch::wheel==0.43.0=pyhd8ed1ab_1[md5=0b5293a157c2b5cd513dd1b03d8d3aae]
   - conda-forge/linux-64::cffi==1.16.0=py310h2fee648_0[md5=45846a970e71ac98fd327da5d40a0a2c]
   - conda-forge/noarch::jsonpatch==1.33=pyhd8ed1ab_0[md5=bfdb7c5c6ad1077c82a69a8642c87aff]
   - conda-forge/linux-64::libmambapy==1.5.8=py310h39ff949_0[md5=37f8aa15b73c4691eeec15caf45aab25]
   - conda-forge/noarch::pip==24.0=pyhd8ed1ab_0[md5=f586ac1e56c8638b64f9c8122a7b8a67]
   - conda-forge/linux-64::ruamel.yaml==0.18.6=py310h2372a71_0[md5=50b7d9b39099cdbabf65bf27df73a793]
   - conda-forge/noarch::tqdm==4.66.2=pyhd8ed1ab_0[md5=2b8dfb969f984497f3f98409a9545776]
   - conda-forge/noarch::urllib3==2.2.1=pyhd8ed1ab_0[md5=08807a87fa7af10754d46f63b368e016]
   - conda-forge/noarch::requests==2.31.0=pyhd8ed1ab_0[md5=a30144e4156cdbb236f99ebb49828f8b]
   - conda-forge/linux-64::zstandard==0.22.0=py310h1275a96_0[md5=54698ba13cd3494547b289cd86a2176a]
   - conda-forge/noarch::conda-package-streaming==0.9.0=pyhd8ed1ab_0[md5=38253361efb303deead3eab39ae9269b]
   - conda-forge/noarch::conda-package-handling==2.2.0=pyh38be061_0[md5=8a3ae7f6318376aa08ea753367bb7dd6]
   - conda-forge/linux-64::conda==24.3.0=py310hff52083_0[md5=4187d17c4b75d8f7757820e835c507c9]
   - conda-forge/noarch::conda-libmamba-solver==24.1.0=pyhd8ed1ab_0[md5=304dc78ad6e52e0fd663df1d484c1531]
   - conda-forge/linux-64::mamba==1.5.8=py310h51d5547_0[md5=3b335eaa4894cbb5379a75f83a4d6b40]


  Package                         Version  Build               Channel         Size
─────────────────────────────────────────────────────────────────────────────────────
  Install:
─────────────────────────────────────────────────────────────────────────────────────

  + _libgcc_mutex                     0.1  conda_forge         conda-forge         
  + ca-certificates              2024.2.2  hbcca054_0          conda-forge         
  + ld_impl_linux-64                 2.40  h41732ed_0          conda-forge         
  + libstdcxx-ng                   13.2.0  h7e041cc_5          conda-forge         
  + pybind11-abi                        4  hd8ed1ab_3          conda-forge         
  + python_abi                       3.10  4_cp310             conda-forge         
  + tzdata                          2024a  h0c530f3_0          conda-forge         
  + libgomp                        13.2.0  h807b86a_5          conda-forge         
  + _openmp_mutex                     4.5  2_gnu               conda-forge         
  + libgcc-ng                      13.2.0  h807b86a_5          conda-forge         
  + bzip2                           1.0.8  hd590300_5          conda-forge         
  + c-ares                         1.28.1  hd590300_0          conda-forge         
  + fmt                            10.2.1  h00ab1b0_0          conda-forge         
  + icu                              73.2  h59595ed_0          conda-forge         
  + keyutils                        1.6.1  h166bdaf_0          conda-forge         
  + libev                            4.33  hd590300_2          conda-forge         
  + libffi                          3.4.2  h7f98852_5          conda-forge         
  + libiconv                         1.17  hd590300_2          conda-forge         
  + libnsl                          2.0.1  hd590300_0          conda-forge         
  + libuuid                        2.38.1  h0b41bf4_0          conda-forge         
  + libxcrypt                      4.4.36  hd590300_1          conda-forge         
  + libzlib                        1.2.13  hd590300_5          conda-forge         
  + lz4-c                           1.9.4  hcb278e6_0          conda-forge         
  + lzo                              2.10  h516909a_1000       conda-forge         
  + ncurses                  6.4.20240210  h59595ed_0          conda-forge         
  + openssl                         3.2.1  hd590300_1          conda-forge         
  + reproc                   14.2.4.post0  hd590300_1          conda-forge         
  + xz                              5.2.6  h166bdaf_0          conda-forge         
  + yaml-cpp                        0.8.0  h59595ed_0          conda-forge         
  + libedit                  3.1.20191231  he28a2e2_2          conda-forge         
  + libnghttp2                     1.58.0  h47da74e_1          conda-forge         
  + libsolv                        0.7.28  hfc55251_2          conda-forge         
  + libsqlite                      3.45.2  h2797004_0          conda-forge         
  + libssh2                        1.11.0  h0841786_0          conda-forge         
  + libxml2                        2.12.6  h232c23b_1          conda-forge         
  + readline                          8.2  h8228510_1          conda-forge         
  + reproc-cpp               14.2.4.post0  h59595ed_1          conda-forge         
  + tk                             8.6.13  noxft_h4845f30_101  conda-forge         
  + zstd                            1.5.5  hfc55251_0          conda-forge         
  + krb5                           1.21.2  h659d440_0          conda-forge         
  + libarchive                      3.7.2  h2aa1ff5_1          conda-forge         
  + python                        3.10.14  hd12c33a_0_cpython  conda-forge         
  + libcurl                         8.7.1  hca28451_0          conda-forge         
  + menuinst                        2.0.2  py310hff52083_0     conda-forge         
  + archspec                        0.2.3  pyhd8ed1ab_0        conda-forge         
  + boltons                        24.0.0  pyhd8ed1ab_0        conda-forge         
  + brotli-python                   1.1.0  py310hc6cd4ac_1     conda-forge         
  + certifi                      2024.2.2  pyhd8ed1ab_0        conda-forge         
  + charset-normalizer              3.3.2  pyhd8ed1ab_0        conda-forge         
  + colorama                        0.4.6  pyhd8ed1ab_0        conda-forge         
  + distro                          1.9.0  pyhd8ed1ab_0        conda-forge         
  + idna                              3.6  pyhd8ed1ab_0        conda-forge         
  + jsonpointer                       2.4  py310hff52083_3     conda-forge         
  + libmamba                        1.5.8  had39da4_0          conda-forge         
  + packaging                        24.0  pyhd8ed1ab_0        conda-forge         
  + platformdirs                    4.2.0  pyhd8ed1ab_0        conda-forge         
  + pluggy                          1.4.0  pyhd8ed1ab_0        conda-forge         
  + pycosat                         0.6.6  py310h2372a71_0     conda-forge         
  + pycparser                        2.22  pyhd8ed1ab_0        conda-forge         
  + pysocks                         1.7.1  pyha2e5f31_6        conda-forge         
  + ruamel.yaml.clib                0.2.8  py310h2372a71_0     conda-forge         
  + setuptools                     69.5.1  pyhd8ed1ab_0        conda-forge         
  + truststore                      0.8.0  pyhd8ed1ab_0        conda-forge         
  + wheel                          0.43.0  pyhd8ed1ab_1        conda-forge         
  + cffi                           1.16.0  py310h2fee648_0     conda-forge         
  + jsonpatch                        1.33  pyhd8ed1ab_0        conda-forge         
  + libmambapy                      1.5.8  py310h39ff949_0     conda-forge         
  + pip                              24.0  pyhd8ed1ab_0        conda-forge         
  + ruamel.yaml                    0.18.6  py310h2372a71_0     conda-forge         
  + tqdm                           4.66.2  pyhd8ed1ab_0        conda-forge         
  + urllib3                         2.2.1  pyhd8ed1ab_0        conda-forge         
  + requests                       2.31.0  pyhd8ed1ab_0        conda-forge         
  + zstandard                      0.22.0  py310h1275a96_0     conda-forge         
  + conda-package-streaming         0.9.0  pyhd8ed1ab_0        conda-forge         
  + conda-package-handling          2.2.0  pyh38be061_0        conda-forge         
  + conda                          24.3.0  py310hff52083_0     conda-forge         
  + conda-libmamba-solver          24.1.0  pyhd8ed1ab_0        conda-forge         
  + mamba                           1.5.8  py310h51d5547_0     conda-forge         

  Summary:

  Install: 78 packages

  Total download: 0 B

─────────────────────────────────────────────────────────────────────────────────────

Transaction starting
Linking _libgcc_mutex-0.1-conda_forge
Linking ca-certificates-2024.2.2-hbcca054_0
Linking ld_impl_linux-64-2.40-h41732ed_0
Linking libstdcxx-ng-13.2.0-h7e041cc_5
Linking pybind11-abi-4-hd8ed1ab_3
Linking python_abi-3.10-4_cp310
Linking tzdata-2024a-h0c530f3_0
Linking libgomp-13.2.0-h807b86a_5
Linking _openmp_mutex-4.5-2_gnu
Linking libgcc-ng-13.2.0-h807b86a_5
Linking bzip2-1.0.8-hd590300_5
Linking c-ares-1.28.1-hd590300_0
Linking fmt-10.2.1-h00ab1b0_0
Linking icu-73.2-h59595ed_0
Linking keyutils-1.6.1-h166bdaf_0
Linking libev-4.33-hd590300_2
Linking libffi-3.4.2-h7f98852_5
Linking libiconv-1.17-hd590300_2
Linking libnsl-2.0.1-hd590300_0
Linking libuuid-2.38.1-h0b41bf4_0
Linking libxcrypt-4.4.36-hd590300_1
Linking libzlib-1.2.13-hd590300_5
Linking lz4-c-1.9.4-hcb278e6_0
Linking lzo-2.10-h516909a_1000
Linking ncurses-6.4.20240210-h59595ed_0
Linking openssl-3.2.1-hd590300_1
Linking reproc-14.2.4.post0-hd590300_1
Linking xz-5.2.6-h166bdaf_0
Linking yaml-cpp-0.8.0-h59595ed_0
Linking libedit-3.1.20191231-he28a2e2_2
Linking libnghttp2-1.58.0-h47da74e_1
Linking libsolv-0.7.28-hfc55251_2
Linking libsqlite-3.45.2-h2797004_0
Linking libssh2-1.11.0-h0841786_0
Linking libxml2-2.12.6-h232c23b_1
Linking readline-8.2-h8228510_1
Linking reproc-cpp-14.2.4.post0-h59595ed_1
Linking tk-8.6.13-noxft_h4845f30_101
Linking zstd-1.5.5-hfc55251_0
Linking krb5-1.21.2-h659d440_0
Linking libarchive-3.7.2-h2aa1ff5_1
Linking python-3.10.14-hd12c33a_0_cpython
Linking libcurl-8.7.1-hca28451_0
Linking menuinst-2.0.2-py310hff52083_0
Linking archspec-0.2.3-pyhd8ed1ab_0
Linking boltons-24.0.0-pyhd8ed1ab_0
Linking brotli-python-1.1.0-py310hc6cd4ac_1
Linking certifi-2024.2.2-pyhd8ed1ab_0
Linking charset-normalizer-3.3.2-pyhd8ed1ab_0
Linking colorama-0.4.6-pyhd8ed1ab_0
Linking distro-1.9.0-pyhd8ed1ab_0
Linking idna-3.6-pyhd8ed1ab_0
Linking jsonpointer-2.4-py310hff52083_3
Linking libmamba-1.5.8-had39da4_0
Linking packaging-24.0-pyhd8ed1ab_0
Linking platformdirs-4.2.0-pyhd8ed1ab_0
Linking pluggy-1.4.0-pyhd8ed1ab_0
Linking pycosat-0.6.6-py310h2372a71_0
Linking pycparser-2.22-pyhd8ed1ab_0
Linking pysocks-1.7.1-pyha2e5f31_6
Linking ruamel.yaml.clib-0.2.8-py310h2372a71_0
Linking setuptools-69.5.1-pyhd8ed1ab_0
Linking truststore-0.8.0-pyhd8ed1ab_0
Linking wheel-0.43.0-pyhd8ed1ab_1
Linking cffi-1.16.0-py310h2fee648_0
Linking jsonpatch-1.33-pyhd8ed1ab_0
Linking libmambapy-1.5.8-py310h39ff949_0
Linking pip-24.0-pyhd8ed1ab_0
Linking ruamel.yaml-0.18.6-py310h2372a71_0
Linking tqdm-4.66.2-pyhd8ed1ab_0
Linking urllib3-2.2.1-pyhd8ed1ab_0
Linking requests-2.31.0-pyhd8ed1ab_0
Linking zstandard-0.22.0-py310h1275a96_0
Linking conda-package-streaming-0.9.0-pyhd8ed1ab_0
Linking conda-package-handling-2.2.0-pyh38be061_0
Linking conda-24.3.0-py310hff52083_0
Linking conda-libmamba-solver-24.1.0-pyhd8ed1ab_0
Linking mamba-1.5.8-py310h51d5547_0

Transaction finished

To activate this environment, use:

    micromamba activate /home/u19/caparicio/miniforge3

Or to execute a single command in this environment, use:

    micromamba run -p /home/u19/caparicio/miniforge3 mycommand

installation finished.
WARNING:
    You currently have a PYTHONPATH environment variable set. This may cause
    unexpected behavior when running the Python interpreter in Miniforge3.
    For best results, please verify that your PYTHONPATH only points to
    directories of packages that are compatible with the Python interpreter
    in Miniforge3: /home/u19/caparicio/miniforge3
Do you wish to update your shell profile to automatically initialize conda?
This will activate conda on startup and change the command prompt when activated.
If you'd prefer that conda's base environment not be activated on startup,
   run the following command when conda is activated:

conda config --set auto_activate_base false

You can undo this by running `conda init --reverse $SHELL`? [yes|no]
[no] >>> no

You have chosen to not have conda modify your shell scripts at all.
To activate conda's base environment in your current shell session:

eval "$(/home/u19/caparicio/miniforge3/bin/conda shell.YOUR_SHELL_NAME hook)" 

To install conda's shell functions for easier access, first activate, then:

conda init

Thank you for installing Miniforge3!
```
